distributed/tests/test_scheduler.py

```diff
@@ -126,6 +126,22 @@ async def test_decide_worker_with_restrictions(client, s, a, b, c):
     assert x.key in a.data or x.key in b.data
 
 
+@gen_cluster(
+    client=True,
+    nthreads=[("127.0.0.1", 1)] * 3,
+    config={"distributed.scheduler.work-stealing": False},
+)
+async def test_decide_worker_select_candidate_holding_no_deps(client, s, a, b, c):
+    root = await client.scatter(1)
+    assert sum(root.key in worker.data for worker in [a, b, c]) == 1
+
+    tasks = client.map(inc, [root] * 6, pure=False)
+    await wait(tasks)
+
+    assert all(root.key in worker.data for worker in [a, b, c])
+    assert len(a.data) == len(b.data) == len(c.data) == 3
```

Comment:
Actually, now that I look at this again, maybe we do need the sleep.

If we're scheduling `inc`s then it's not unreasonable to schedule six of these things on one worker.  That's where the data is.   This computation is very cheap.  But in contrast, the dependency is very small.  We may need the scheduler to learn that `slowinc` is slow, and so heavy relative to the `int` dependency.  

Maybe 

```python
await client.submit(slowinc, 10, delay=0.1)  # learn that slowinc is slow
root = await client.scatter(1)

futures = client.submit(slowinc, [root] * 6, delay=0.1, pure=False)
await wait(futures)
```

This way we're confident that the computational cost of the future will far outweigh the communciation cost, but things are still fast-ish.

Reply:
Yeah, that's why I originally went with `sleep`. It also _feels_ more like the case we're trying to test for. Though the test still passes with 6d91816 (and note that that test fails on main), so I'm not sure what we're thinking about wrong?

Reply:
Currently 28 bytes / 100MB/s is just still smaller than however fast inc ran in (a few microseconds probably). It would be safer to bump this up to milliseconds though.  

gen_cluster tests can run in less than a second, but more than 100ms.  So sleeps of around that 100ms time are more welcome than 1s sleeps.

Reply:
The test is passing as-is; do you still think I should change it back to `sleep`?

Reply:
It would become a problem if the `inc` ran in `28 bytes / 100MB/s = 0.2us` or less.  Inc does actually run this fast.

```python
In [1]: 28 / 100e6
Out[1]: 2.8e-07

In [2]: def inc(x):
   ...:     return x + 1
   ...: 

In [3]: %timeit inc(1)
64.2 ns ± 0.914 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

But presumably the infrastructure around the inc is taking longer.  If our timing infrsatructure got much faster then this might result in intermittent errors.  It *probably* doesn't matter at this scale, but it's probably a good idea if it takes only a few minutes.

We might also introduce a penalty of something like 1ms for *any* communication (I'm surprised that we don't do this today actually) which might tip the scales in the future.

Reply:
I'll switch back to a short sleep for now.

> We might also introduce a penalty of something like 1ms for _any_ communication

That seems very worth doing.

Reply:
We maybe used to do this?  It might be worth checking logs.  The `worker_objective` function is pretty clean if you wanted to add this.  We should probably wait until we find a case where it comes up though in an impactful way just to keep from cluttering things.

Reply:
Yeah I think it should be a separate PR for cleanliness